### PR TITLE
Canceled ride feature

### PR DESCRIPTION
**Pull request by gandhinikhil**
Adding new feature for Cancelling ride for user.

### CODE REVIEW

File: src/main/java/com/example/cabbooking/servicetest/RideService.java [hunk 1]
```diff
@@ -28,42 +36,54 @@ public List<RideDetailsVO> findRide(String username, LocationDetailsVO source, L
...
-                return Double.compare(distanceToO1,distanceToO2);
+                double distanceToO1 = calculateDistance(sources, o1.getDriver().getLocation());
+                double distanceToO2 = calculateDistance(sources, o2.getDriver().getLocation());
+                return Double.compare(distanceToO1, distanceToO2);
             }
         });
     }
 
+    public void cancelRide(String userName, String driverName) {
+        Ride ride = rideRepository.findByUserNameAndDriverName(userName, driverName)
```
Comment (sanskrati3) [inline src/main/java/com/example/cabbooking/servicetest/RideService.java:83]: add proper null checks what if userName and driverName does nit exists?

File: src/main/java/com/example/cabbooking/servicetest/RideService.java [hunk 2]
```diff
@@ -28,42 +36,54 @@ public List<RideDetailsVO> findRide(String username, LocationDetailsVO source, L
...
-                double distanceToO2 = calculateDistance(sources,o2.getDriver().getLocation());
-                return Double.compare(distanceToO1,distanceToO2);
+                double distanceToO1 = calculateDistance(sources, o1.getDriver().getLocation());
+                double distanceToO2 = calculateDistance(sources, o2.getDriver().getLocation());
+                return Double.compare(distanceToO1, distanceToO2);
             }
         });
     }
 
+    public void cancelRide(String userName, String driverName) {
```
Comment (sanskrati3) [inline src/main/java/com/example/cabbooking/servicetest/RideService.java:78]: is it possible for user to hit this API directly before even requesting for a ride ?this should follow a order right ? first login -> thn select address -> then see fare -> then book -> then cancel how are you maintaining that ?

File: src/main/java/com/example/cabbooking/servicetest/RideService.java [hunk 3]
```diff
@@ -28,42 +36,54 @@ public List<RideDetailsVO> findRide(String username, LocationDetailsVO source, L
...
+        Ride ride = rideRepository.findByUserNameAndDriverName(userName, driverName)
+                .orElseThrow(() -> new RideNotFoundException("Ride not found for user: " + userName + " with driver: " + driverName));
+
+        if (ride.isCanceled()) {
+            throw new AlreadyCanceledException("Ride already canceled for user: " + userName + " with driver: " + driverName);
```
Comment (sanskrati3) [inline src/main/java/com/example/cabbooking/servicetest/RideService.java:87]: please check everytime the prerequisits to for cancaling a ride

File: src/main/java/com/example/cabbooking/servicetest/RideService.java (hunk 1 above)
Comment (gandhinikhil) [reply src/main/java/com/example/cabbooking/servicetest/RideService.java:83]: Check Null updates

File: src/main/java/com/example/cabbooking/servicetest/RideService.java (hunk 2 above)
Comment (gandhinikhil) [reply src/main/java/com/example/cabbooking/servicetest/RideService.java:78]: this feature will updated but once user chose ride after then user will able to cancel ride. Next feature will take care of it order.

File: src/main/java/com/example/cabbooking/servicetest/RideService.java (hunk 3 above)
Comment (gandhinikhil) [reply src/main/java/com/example/cabbooking/servicetest/RideService.java:87]: In this case we once we cancel the ride status become true.

**Pull request merged**